### ISSUE TITLE
[Bug]: Qwen3.5-9B hybrid-GDN + dynamic LoRA on H20 produces NaN output ("!" tokens) for long sequences — punica Triton kernel bug

### ISSUE THREAD

**Issue by XinXinShawn** [vllm-project/vllm#52568]
### Your current environment

<details>
<summary>The output of <code>python collect_env.py</code></summary>

```text
Collecting environment information...
==============================
        System Info
==============================
OS                           : Ubuntu 22.04.5 LTS (x86_64)
GCC version                  : (Ubuntu 11.4.0-1ubuntu1~22.04.3) 11.4.0
Clang version                : Could not collect
CMake version                : Could not collect
Libc version                 : glibc-2.35

==============================
       PyTorch Info
==============================
PyTorch version              : 2.13.0+cu130
Is debug build               : False
CUDA used to build PyTorch   : 13.0
ROCM used to build PyTorch   : N/A
XPU used to build PyTorch    : N/A

==============================
      Python Environment
==============================
Python version               : 3.12.13 (main, Mar  4 2026, 09:23:07) [GCC 11.4.0] (64-bit runtime)
Python platform              : Linux-5.10.134-19.2.al8.x86_64-x86_64-with-glibc2.35

==============================
       CUDA / GPU Info
==============================
Is CUDA available            : True
CUDA runtime version         : 13.0.88
CUDA_MODULE_LOADING set to   :
GPU models and configuration : GPU 0: NVIDIA H20
Nvidia driver version        : 580.126.09
cuDNN version                : Could not collect
HIP runtime version          : N/A
MIOpen runtime version       : N/A
Is XNNPACK available         : False

==============================
          CPU Info
==============================
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      52 bits physical, 57 bits virtual
Byte Order:                         Little Endian
CPU(s):                             24
On-line CPU(s) list:                0-23
Vendor ID:                          GenuineIntel
Model name:                         Intel(R) Xeon(R) Platinum 8469C
CPU family:                         6
Model:                              143
Thread(s) per core:                 2
Core(s) per socket:                 12
Socket(s):                          1
Stepping:                           8
CPU max MHz:                        3800.0000
CPU min MHz:                        800.0000
BogoMIPS:                           5200.00
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq monitor ssse3 fma cx16 pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch cpuid_fault invpcid_single ibrs_enhanced fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves avx_vnni avx512_bf16 wbnoinvd ida arat hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq rdpid bus_lock_detect cldemote movdiri movdir64b enqcmd fsrm uintr md_clear serialize tsxldtrk amx_bf16 avx512_fp16 amx_tile amx_int8 arch_capabilities
Hypervisor vendor:                  KVM
Virtualization type:                full
L1d cache:                          576 KiB (12 instances)
L1i cache:                          384 KiB (12 instances)
L2 cache:                           24 MiB (12 instances)
L3 cache:                           97.5 MiB (1 instance)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0-23
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Not affected
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Vulnerable
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Enhanced / Automatic IBRS, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

==============================
Versions of relevant libraries
==============================
[pip3] flashinfer-python==0.6.16.post3
[pip3] nccl4py==0.3.1
[pip3] numpy==2.2.6
[pip3] nvidia-cublas==13.1.1.3
[pip3] nvidia-cuda-cccl==13.3.3.4.1
[pip3] nvidia-cuda-crt==13.3.73
[pip3] nvidia-cuda-cupti==13.0.85
[pip3] nvidia-cuda-nvcc==13.3.73
[pip3] nvidia-cuda-nvdisasm==13.3.73
[pip3] nvidia-cuda-nvrtc==13.0.88
[pip3] nvidia-cuda-runtime==13.0.96
[pip3] nvidia-cudnn-cu13==9.20.0.48
[pip3] nvidia-cudnn-frontend==1.27.0
[pip3] nvidia-cufft==12.0.0.61
[pip3] nvidia-cufile==1.15.1.6
[pip3] nvidia-curand==10.4.0.35
[pip3] nvidia-cusolver==12.0.4.66
[pip3] nvidia-cusparse==12.6.3.3
[pip3] nvidia-cusparselt-cu13==0.8.1
[pip3] nvidia-cutlass-dsl==4.6.0
[pip3] nvidia-cutlass-dsl-libs-base==4.6.0
[pip3] nvidia-cutlass-dsl-libs-core==4.6.0
[pip3] nvidia-cutlass-dsl-libs-cu12==4.6.0
[pip3] nvidia-cutlass-dsl-libs-cu13==4.6.0
[pip3] nvidia-ml-py==13.610.43
[pip3] nvidia-nccl-cu13==2.30.7
[pip3] nvidia-nvjitlink==13.3.33
[pip3] nvidia-nvshmem-cu13==3.4.5
[pip3] nvidia-nvtx==13.0.85
[pip3] nvidia-nvvm==13.3.73
[pip3] pyzmq==27.1.0
[pip3] tokenspeed-triton==3.8.10.post20260721
[pip3] torch==2.13.0+cu130
[pip3] torch_c_dlpack_ext==0.1.5
[pip3] torchaudio==2.11.0+cu130
[pip3] torchcodec==0.15.0+cu130
[pip3] torchvision==0.28.0+cu130
[pip3] transformers==5.15.0
[pip3] triton==3.7.1
[conda] Could not collect

==============================
         vLLM Info
==============================
ROCM Version                 : Could not collect
vLLM Version                 : 0.27.0
vLLM Build Flags:
  CUDA Archs: 7.5 8.0 8.6 8.9 9.0 10.0 12.0; ROCm: Disabled; XPU: Disabled
GPU Topology:
        GPU0    CPU Affinity    NUMA Affinity   GPU NUMA ID
GPU0     X      0-23    0               N/A

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

==============================
     Environment Variables
==============================
NVIDIA_VISIBLE_DEVICES=all
VLLM_BUILD_URL=https://buildkite.com/vllm/release-v2/builds/4962
NVIDIA_REQUIRE_CUDA=cuda>=13.0 brand=unknown,driver>=535,driver<536 brand=grid,driver>=535,driver<536 brand=tesla,driver>=535,driver<536 brand=nvidia,driver>=535,driver<536 brand=quadro,driver>=535,driver<536 brand=quadrortx,driver>=535,driver<536 brand=nvidiartx,driver>=535,driver<536 brand=vapps,driver>=535,driver<536 brand=vpc,driver>=535,driver<536 brand=vcs,driver>=535,driver<536 brand=vws,driver>=535,driver<536 brand=cloudgaming,driver>=535,driver<536 brand=unknown,driver>=550,driver<551 brand=grid,driver>=550,driver<551 brand=tesla,driver>=550,driver<551 brand=nvidia,driver>=550,driver<551 brand=quadro,driver>=550,driver<551 brand=quadrortx,driver>=550,driver<551 brand=nvidiartx,driver>=550,driver<551 brand=vapps,driver>=550,driver<551 brand=vpc,driver>=550,driver<551 brand=vcs,driver>=550,driver<551 brand=vws,driver>=550,driver<551 brand=cloudgaming,driver>=550,driver<551 brand=unknown,driver>=565,driver<566 brand=grid,driver>=565,driver<566 brand=tesla,driver>=565,driver<566 brand=nvidia,driver>=565,driver<566 brand=quadro,driver>=565,driver<566 brand=quadrortx,driver>=565,driver<566 brand=nvidiartx,driver>=565,driver<566 brand=vapps,driver>=565,driver<566 brand=vpc,driver>=565,driver<566 brand=vcs,driver>=565,driver<566 brand=vws,driver>=565,driver<566 brand=cloudgaming,driver>=565,driver<566 brand=unknown,driver>=570,driver<571 brand=grid,driver>=570,driver<571 brand=tesla,driver>=570,driver<571 brand=nvidia,driver>=570,driver<571 brand=quadro,driver>=570,driver<571 brand=quadrortx,driver>=570,driver<571 brand=nvidiartx,driver>=570,driver<571 brand=vapps,driver>=570,driver<571 brand=vpc,driver>=570,driver<571 brand=vcs,driver>=570,driver<571 brand=vws,driver>=570,driver<571 brand=cloudgaming,driver>=570,driver<571 brand=unknown,driver>=575,driver<576 brand=grid,driver>=575,driver<576 brand=tesla,driver>=575,driver<576 brand=nvidia,driver>=575,driver<576 brand=quadro,driver>=575,driver<576 brand=quadrortx,driver>=575,driver<576 brand=nvidiartx,driver>=575,driver<576 brand=vapps,driver>=575,driver<576 brand=vpc,driver>=575,driver<576 brand=vcs,driver>=575,driver<576 brand=vws,driver>=575,driver<576 brand=cloudgaming,driver>=575,driver<576
TORCH_CUDA_ARCH_LIST=7.5 8.0 8.6 8.9 9.0 10.0 12.0
NVIDIA_DRIVER_CAPABILITIES=compute,utility
VLLM_IMAGE_TAG=vllm/vllm-openai:v0.27.0
VLLM_USAGE_SOURCE=production-docker-image
CUDA_VERSION=13.0.2
VLLM_CONFIGURE_LOGGING=1
VLLM_ENABLE_CUDA_COMPATIBILITY=0
PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True
VLLM_BUILD_PIPELINE=019d130e-464e-4ff7-b84b-492992c0c06b
LD_LIBRARY_PATH=/usr/local/nvidia/lib64:/usr/local/cuda/lib64:/usr/local/nvidia/lib:/usr/local/nvidia/lib64:/usr/local/cuda/lib64
VLLM_BUILD_COMMIT=4bdc8a788d2e2ce9165d552b3d4d8b72604626bf
PYTORCH_NVML_BASED_CUDA_CHECK=1
TORCHINDUCTOR_COMPILE_THREADS=1
TORCHINDUCTOR_CACHE_DIR=/tmp/torchinductor_root
```

</details>

### my environment

- GPU: NVIDIA H20 (Hopper, SM90)
- Model: Qwen/Qwen3.5-9B (hybrid-GDN: 24 linear_attn layers + 8 self_attn layers)
- LoRA adapter: trained with Unsloth, r=64, target_modules includes q_proj/k_proj/v_proj/o_proj (self_attn), in_proj_qkv/in_proj_z/out_proj (linear_attn), gate_proj/up_proj/down_proj (MLP)
- Docker image: vllm/vllm-openai:v0.27.0

### Startup command

```bash
python -m vllm.entrypoints.openai.api_server \
  --model /model \
  --served-model-name Qwen3.5-9B-Base \
  --trust-remote-code \
  --language-model-only \
  --enforce-eager \
  --dtype bfloat16 \
  --max-model-len 65536 \
  --gpu-memory-utilization 0.8 \
  --enable-lora \
  --max-lora-rank 64 \
  --max-loras 6 \
  --lora-modules 0814=/loras/qwen3.5-9b-nlu-lora-0814
### Reproduction steps

1. Start vLLM with `--enable-lora` and a Qwen3.5-9B LoRA adapter
2. Send a short prompt (< ~400 tokens) with the LoRA adapter → output is normal
3. Send a long prompt (> ~400 tokens) with the LoRA adapter → output is all `!` characters
### Expected behavior

LoRA adapter should produce correct output for both short and long sequences, matching:

- HuggingFace PEFT `PeftModel.from_pretrained()` inference (works correctly)
- Merged model loaded directly in vLLM (works correctly)

### Actual behavior

- **Short sequences (< ~400 tokens)**: normal output ✅
- **Long sequences (> ~400 tokens)**: all output tokens are `!` (token id 0) ❌
- **Base model (no LoRA)**: normal output ✅
- **Merged model in vLLM**: normal output ✅
- **HF PEFT inference**: normal output ✅

The `!` output is caused by NaN in logits: `argmax(NaN) → token 0 → "!"`.

### Root cause analysis

This is the same issue described in this CSDN blog post (Chinese):
https://blog.csdn.net/qq_26624329/article/details/162686302
### What I've tried

1. ✅ Changed `torch.empty` → `torch.zeros` in `add_lora_linear` → **still produces `!`** (the bug is deeper than buffer init, likely in `lora_shrink`/`lora_expand` Triton kernels themselves)
2. ✅ `--enforce-eager` → no effect
3. ✅ `--language-model-only` → no effect
4. ✅ Various `--gpu-memory-utilization` values → no effect
5. ✅ Disabled prefix caching / chunked prefill → no effect
6. ✅ Different sampling parameters (temperature, top_p) → no effect
7. ✅ Retrained LoRA with full linear_attn coverage (in_proj_qkv, in_proj_z, out_proj) → no effect

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by Tejas-Raj01** [comment]
Hi @jeejeelee, since I've already tracked down the root cause and tested the patch, would you mind assigning this issue to me so I can open a PR for it? I'd be happy to handle the implementation, add the necessary tests, and push it through CI. However, if you are already working on a broader fix, I am entirely open to collaborating or submitting this patch as a starting point. Let me know what works best for you!

Here is the root cause analysis and the proposed patch that resolves the issue.

### Root Cause Analysis

There are two overlapping issues causing the `!` token outputs on sequences longer than ~400 tokens:

**1. Catastrophic Accumulator Overflow in Triton**
As sequence lengths grow, the variance and magnitude of the hidden states increase. In the Punica Triton kernels (`lora_shrink_op.py` and `lora_expand_op.py`), the dot product accumulator implicitly inherits the input data type (`bfloat16`). When multiplying and accumulating these larger vectors across the massive hidden dimensions of Qwen3.5's `hybrid-GDN`, the running sum exceeds the 16-bit dynamic range. This results in an `Inf`, cascading into a `NaN`, which collapses the `argmax` into token `0` (`!`).

**2. The Persistent Workspace Bypass**
The previous attempt to change `torch.empty` to `torch.zeros` in `add_lora_linear` fails to fix the issue. This happens because vLLM pre-allocates a persistent workspace buffer at the `WorkerManager` level. Because the `buffer` is passed into `add_lora_linear` as an argument, the `if buffer is None:` block is bypassed entirely. The kernels pull from the pre-allocated memory pool, which often contains garbage NaNs from previous GPU operations.

---

### The Fix

We need to patch both the Triton kernels and the Python wrapper.

**Step 1: Force `float32` Accumulation in the Triton Kernels**
In `vllm/lora/ops/triton_ops/lora_shrink_op.py` and `lora_expand_op.py`, force the accumulator to use `float32` and only cast back at the store operation:

```python
# 1. Update initialization
acc = tl.zeros((BLOCK_M, BLOCK_N), dtype=tl.float32)

# 2. Update dot product
acc += tl.dot(a, b, out_dtype=tl.float32)

# 3. Cast back on store
tl.store(out_ptr + ..., acc.to(inputs.dtype), mask=...)
```

**Step 2: Aggressively Zero the Workspace Buffer**
In the wrapper (e.g., `vllm/lora/punica_wrapper/punica_gpu.py`), we must guarantee the injected buffer is cleared:

```python
    def add_lora_linear(
        self,
        y: torch.Tensor,
        x: torch.Tensor,
        lora_a_stacked: tuple[torch.Tensor, ...],
        lora_b_stacked: tuple[torch.Tensor, ...],
        scale: float,
        output_slices: tuple[int, ...],
        *,
        buffer: tuple[torch.Tensor, ...] | None = None,
        **kwargs,
    ) -> torch.Tensor | None:
        
        if buffer is None:
            buffer = tuple(
                torch.zeros((x.shape[0], lora_b_stacked[i].shape[1]),
                            dtype=x.dtype,
                            device=x.device) for i in range(len(output_slices))
            )
        else:
            # Force zero the injected persistent workspace buffer
            for b in buffer:
                b.zero_()
                
        # ... proceed to self.add_shrink and self.add_expand
```

*Note: Anyone applying this locally needs to run `rm -rf ~/.triton/cache` to clear the old binaries before testing.*

**Comment by WudiJoey** [comment]
@Tejas-Raj01  I'm also encountering this issue. Has this bug been fixed in the latest version 0.28.0？

**Comment by XinXinShawn** [comment]
I've tested on vllm 0.28.0, this bug is not fixed. It still outputs  !  when token length gets long.



---- Replied Message ----
| From | ***@***.***> |
| Date | 08/31/2026 20:50 |
| To | ***@***.***> |
| Cc | ***@***.***>***@***.***> |
| Subject | Re: [vllm-project/vllm] [Bug]: Qwen3.5-9B hybrid-GDN + dynamic LoRA on H20 produces NaN output ("!" tokens) for long sequences — punica Triton kernel bug (Issue #52568) |
WudiJoey left a comment (vllm-project/vllm#52568)

@Tejas-Raj01 I'm also encountering this issue. Has this bug been fixed in the latest version 0.28.0？

—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications, keep track of coding agent tasks and review pull requests on the go with GitHub Mobile for iOS and Android. Download it today!

**Comment by WudiJoey** [comment]
您好，您的来信我已经收到，会尽快回复。
